src/core/labeling/qgspallabeling.cpp

```diff
@@ -2373,119 +2373,159 @@ std::unique_ptr<QgsLabelFeature> QgsPalLayerSettings::registerFeatureWithDetails
                 }
               }
             }
+          }
+        }
 
-            if ( dataDefinedRotation )
-            {
-              //adjust xdiff and ydiff because the hali/vali point needs to be the rotation center
-              double xd = xdiff * std::cos( angle ) - ydiff * std::sin( angle );
-              double yd = xdiff * std::sin( angle ) + ydiff * std::cos( angle );
-              xdiff = xd;
-              ydiff = yd;
-            }
-
-            //project xPos and yPos from layer to map CRS, handle rotation
-            QgsGeometry ddPoint( new QgsPoint( xPos, yPos ) );
-            if ( QgsPalLabeling::geometryRequiresPreparation( ddPoint, context, ct ) )
-            {
-              ddPoint = QgsPalLabeling::prepareGeometry( ddPoint, context, ct );
-              if ( const QgsPoint *point = qgsgeometry_cast< const QgsPoint * >( ddPoint.constGet() ) )
-              {
-                xPos = point->x();
-                yPos = point->y();
-                anchorPosition = QgsPointXY( xPos, yPos );
-              }
-              else
-              {
-                QgsMessageLog::logMessage( QObject::tr( "Invalid data defined label position (%1, %2)" ).arg( xPos ).arg( yPos ), QObject::tr( "Labeling" ) );
-                hasDataDefinedPosition = false;
-              }
-            }
-            else
-            {
-              anchorPosition = QgsPointXY( xPos, yPos );
-            }
+        if ( dataDefinedRotation )
+        {
+          //adjust xdiff and ydiff because the hali/vali point needs to be the rotation center
+          double xd = xdiff * std::cos( angle ) - ydiff * std::sin( angle );
+          double yd = xdiff * std::sin( angle ) + ydiff * std::cos( angle );
+          xdiff = xd;
+          ydiff = yd;
+        }
 
-            xPos += xdiff;
-            yPos += ydiff;
+        //project xPos and yPos from layer to map CRS, handle rotation
+        QgsGeometry ddPoint( new QgsPoint( xPos, yPos ) );
+        if ( QgsPalLabeling::geometryRequiresPreparation( ddPoint, context, ct ) )
+        {
+          ddPoint = QgsPalLabeling::prepareGeometry( ddPoint, context, ct );
+          if ( const QgsPoint *point = qgsgeometry_cast< const QgsPoint * >( ddPoint.constGet() ) )
+          {
+            xPos = point->x();
+            yPos = point->y();
+            anchorPosition = QgsPointXY( xPos, yPos );
+          }
+          else
+          {
+            QgsMessageLog::logMessage( QObject::tr( "Invalid data defined label position (%1, %2)" ).arg( xPos ).arg( yPos ), QObject::tr( "Labeling" ) );
+            hasDataDefinedPosition = false;
           }
         }
+        else
+        {
+          anchorPosition = QgsPointXY( xPos, yPos );
+        }
+
+        xPos += xdiff;
+        yPos += ydiff;
+      }
+      else
+      {
+        anchorPosition = QgsPointXY( xPos, yPos );
+
+        // only rotate non-pinned OverPoint placements until other placements are supported in pal::Feature
```

Comment:
this is old --didn't we remove that constraint recently? I suspect there's a merge mistake here which has reintroduced the old code.